include/RAJA/pattern/detail/reduce.hpp

```diff
@@ -0,0 +1,397 @@
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+
+#include "RAJA/util/Operators.hpp"
+#include "RAJA/util/types.hpp"
+
+#define RAJA_DECLARE_REDUCER(OP, POL, COMBINER)               \
+  template <typename T>                                       \
+  class Reduce##OP<POL, T>                                    \
+      : public reduce::detail::BaseReduce##OP<T, COMBINER>    \
+  {                                                           \
+  public:                                                     \
+    using Base = reduce::detail::BaseReduce##OP<T, COMBINER>; \
+    using Base::Base;                                         \
+  };
+
+#define RAJA_DECLARE_ALL_REDUCERS(POL, COMBINER) \
+  RAJA_DECLARE_REDUCER(Sum, POL, COMBINER)       \
+  RAJA_DECLARE_REDUCER(Min, POL, COMBINER)       \
+  RAJA_DECLARE_REDUCER(Max, POL, COMBINER)       \
+  RAJA_DECLARE_REDUCER(MinLoc, POL, COMBINER)    \
+  RAJA_DECLARE_REDUCER(MaxLoc, POL, COMBINER)
+
+namespace RAJA
+{
+
+namespace reduce
+{
+
+#ifdef RAJA_ENABLE_TARGET_OPENMP
+#pragma omp declare target
+#endif
+
+namespace detail
+{
+
+template <typename T, template <typename...> class Op>
+struct op_adapter : private Op<T, T, T> {
+  using operator_type = Op<T, T, T>;
+  static constexpr T identity() { return operator_type::identity(); }
+  RAJA_HOST_DEVICE RAJA_INLINE void operator()(T &val, const T v) const
+  {
+    val = operator_type::operator()(val, v);
+  }
+};
+}  // end detail
+
+template <typename T>
+struct sum : detail::op_adapter<T, RAJA::operators::plus> {
+};
+
+template <typename T>
+struct min : detail::op_adapter<T, RAJA::operators::minimum> {
+};
+
+template <typename T>
+struct max : detail::op_adapter<T, RAJA::operators::maximum> {
+};
+
+#ifdef RAJA_ENABLE_TARGET_OPENMP
+#pragma omp end declare target
+#endif
+
+namespace detail
+{
+
+template <typename T, bool doing_min = true>
+class ValueLoc
+{
+public:
+  T val = doing_min ? operators::limits<T>::max() : operators::limits<T>::min();
```

Comment:
Does this ever pop up anywhere else? This might be a good constexpr function to have separated.